File: 72-edit-distance/edit-distance.go
/*
w1 = horse
w2 = ros

(i, j) - > num operations needed to convert w1[i:] into w2[j:]
if w1 empty -> length of w2 // insert all 
if w2 empty -> length of w1 // remove all

if equal -> (i + 1, j + 1)
else -> MIN OF (i, j + 1) & (i + 1, j) & (i + 1, j + 1)

*/

func minDistance(w1 string, w2 string) int {
    var solve func(i, j int) int
    memo := make([][]int, len(w1) + 1)
    for i := range memo {
        memo[i] = make([]int, len(w2) + 1)
        for j := range memo[i] {
            memo[i][j] = -1
        }
    } 

    solve = func(i, j int) int {
        if memo[i][j] != -1 {
            return memo[i][j]
        }

        // Base Cases
        if i == len(w1) {
            return len(w2[j:])
        }
        if j == len(w2) {
            return len(w1[i:])
        }

        // Recurrence
        if w1[i] == w2[j]{
            memo[i][j] = solve(i + 1, j + 1)
            return memo[i][j]
        }
        memo[i][j] = min(1 + solve(i + 1, j), 1 + solve(i, j + 1), 1 + solve(i + 1, j + 1))
        return memo[i][j]
    }

    return solve(0, 0)
}

File: 72-edit-distance/edit-distance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        n = len(word1)
        m = len(word2)

        if len(word1) == 0 or len(word2) == 0:
            return max(len(word2), len(word1))
        
        memo = {}
        def num_ops(i: int, j: int) -> int:
            if (i, j) in memo:
                return memo[(i, j)]
            # Base Case:
            if i == len(word1):
                return len(word2[j:])
            if j == len(word2):
                return len(word1[i:])
            
            # Options
            if word1[i] == word2[j]:
                memo[(i, j)] = num_ops(i + 1, j + 1)
                return memo[(i, j)] 
            else:
                memo[(i, j)] = min(
                    num_ops(i + 1, j), # delete i or insert in j
                    num_ops(i, j + 1), # delete j or insert in i
                    num_ops(i + 1, j + 1) # replace i or j
                ) + 1
                return memo[(i, j)] 
        
        return num_ops(0, 0)
        